features/mbedtls/targets/TARGET_Silicon_Labs/crypto_management.c

```diff
@@ -0,0 +1,392 @@
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "crypto_management.h"
+#include "em_core.h"
+#include "em_bus.h"
+
+#if defined( CRYPTO_PRESENT )
+#if defined( MBEDTLS_THREADING_C )
+#include "mbedtls/threading.h"
+static mbedtls_threading_mutex_t    crypto_locks[CRYPTO_COUNT];
+static volatile bool                crypto_locks_initialized = false;
+static unsigned int                 acquire_count = 0U;
+#endif /* MBEDTLS_THREADING_C */
+
+#if defined( MBEDTLS_CRYPTO_DEVICE_PREEMPTION )
+/** Preemptable context of CRYPTO hardware module. */
+typedef struct
+{
+    uint32_t CTRL;          /*!< Control Register  */
+    uint32_t WAC;           /*!< Wide Arithmetic Configuration  */
+    uint32_t SEQCTRL;       /*!< Sequence Control  */
+    uint32_t SEQCTRLB;      /*!< Sequence Control B  */
+    uint32_t IEN;           /*!< Interrupt Enable Register  */
+    uint32_t SEQ[5];        /*!< Instruction Sequence registers */
+    CRYPTO_Data260_TypeDef DDATA[5]; /*!< DDATA registers. Covers all data
+                                        registers
+                                        of CRYPTO, including DATA(128 bit),
+                                        DDATA (256bit/260bit),
+                                        QDATA (512bit) registers. */
+    uint32_t regmask;       /*!< Bitmask for which registers to save */
+    uint32_t operands;      /*!< Saving the currently selected operands */
+    bool carry;             /*!< Saving the status of the carry flag */
+} crypto_context_t;
+
+static crypto_context_t preemption_context;
+static bool             is_preempted = false;
+static CORE_DECLARE_IRQ_STATE;
+#endif /* MBEDTLS_CRYPTO_DEVICE_PREEMPTION */
+
+typedef enum
+{
+#if defined( CRYPTO0 )
+    CRYPTO0_ID = 0,
+#elif defined( CRYPTO )
+    CRYPTO_ID = 0,
+#endif
+#if defined( CRYPTO1 )
+    CRYPTO1_ID = 1,
+#endif
+} crypto_instance_number_t;
+
+typedef struct {
+    CRYPTO_TypeDef *device;
+    uint32_t clockMask;
+} crypto_device_t;
+
+static const crypto_device_t crypto_devices[CRYPTO_COUNT] =
+{
+#if defined( CRYPTO0 )
+    {
+        CRYPTO0,
+        _CMU_HFBUSCLKEN0_CRYPTO0_SHIFT
+    },
+#elif defined( CRYPTO )
+    {
+        CRYPTO,
+        _CMU_HFBUSCLKEN0_CRYPTO_SHIFT
+    },
+#endif
+#if defined( CRYPTO1 )
+    {
+        CRYPTO1,
+        _CMU_HFBUSCLKEN0_CRYPTO1_SHIFT
+    },
+#endif
+};
+
+static inline int crypto_management_index_by_device( CRYPTO_TypeDef *device )
+{
+#if defined( CRYPTO0 )
+    if ( device == CRYPTO0 ) return 0;
+#elif defined( CRYPTO )
+    if ( device == CRYPTO ) return 0;
```

Comment:
Yes. On early devices we defined CRYPTO, when there was only one instance. On the next series (xG12x), we added another instance and defined CRYPTO0 and CRYPTO1. For backwards compatibility, CRYPTO is defined as CRYPTO0 on the new devices, but the old devices don't have a definition for CRYPTO0.